# Web Scraping Lab

You will find in this notebook some scrapy exercises to practise your scraping skills.

**Tips:**

- Check the response status code for each request to ensure you have obtained the intended contennt.
- Print the response text in each request to understand the kind of info you are getting and its format.
- Check for patterns in the response text to extract the data/info requested in each question.
- Visit each url and take a look at its source through Chrome DevTools. You'll need to identify the html tags, special class names etc. used for the html content you are expected to extract.

- [Requests library](http://docs.python-requests.org/en/master/#the-user-guide) documentation 
- [Beautiful Soup Doc](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
- [Urllib](https://docs.python.org/3/library/urllib.html#module-urllib)
- [re lib](https://docs.python.org/3/library/re.html)
- [lxml lib](https://lxml.de/)
- [Scrapy](https://scrapy.org/)
- [List of HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
- [HTML basics](http://www.simplehtmlguide.com/cheatsheet.php)
- [CSS basics](https://www.cssbasics.com/#page_start)

#### Below are the libraries and modules you may need. `requests`,  `BeautifulSoup` and `pandas` are imported for you. If you prefer to use additional libraries feel free to uncomment them.

In [1]:
import requests
from bs4 import BeautifulSoup as bs
import pandas as pd
# from pprint import pprint
from lxml import html
# from lxml.html import fromstring
# import urllib.request
# from urllib.request import urlopen
import random
import re
# import scrapy

#### Download, parse (using BeautifulSoup), and print the content from the Trending Developers page from GitHub:

In [2]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/developers'

In [3]:
#your code

req = requests.get(url).content
req

b'\n\n\n\n\n<!DOCTYPE html>\n<html lang="en">\n  <head>\n    <meta charset="utf-8">\n  <link rel="dns-prefetch" href="https://github.githubassets.com">\n  <link rel="dns-prefetch" href="https://avatars.githubusercontent.com">\n  <link rel="dns-prefetch" href="https://github-cloud.s3.amazonaws.com">\n  <link rel="dns-prefetch" href="https://user-images.githubusercontent.com/">\n\n\n\n  <link crossorigin="anonymous" media="all" integrity="sha512-k9NM/a2xYY6wCRcWG7f3ROm4X5CJNikViGX0N8YIxs6sUYAe/j08/RSHXr3fA9wLIy87AMFCgXm6jbvhZhIXWw==" rel="stylesheet" href="https://github.githubassets.com/assets/frameworks-93d34cfdadb1618eb00917161bb7f744.css" />\n  <link crossorigin="anonymous" media="all" integrity="sha512-ZUf6K+vQqMY+RhVzaRmCy2ePbSZad4TkaGRbd6v5gFt6f9Q/nqjDkBDjQgXNmZw7J9mcYxlsE4fhRw7CTluRow==" rel="stylesheet" href="https://github.githubassets.com/assets/site-6547fa2bebd0a8c63e461573691982cb.css" />\n    <link crossorigin="anonymous" media="all" integrity="sha512-mm2SigzEudA9xS8nyiKvpP

#### Display the names of the trending developers retrieved in the previous step.

Your output should be a Python list of developer names. Each name should not contain any html tag.

**Instructions:**

1. Find out the html tag and class names used for the developer names. You can achieve this using Chrome DevTools.

1. Use BeautifulSoup to extract all the html elements that contain the developer names.

1. Use string manipulation techniques to replace whitespaces and linebreaks (i.e. `\n`) in the *text* of each html element. Use a list to store the clean names.

1. Print the list of names.

Your output should look like below:

```
['trimstray (@trimstray)',
 'joewalnes (JoeWalnes)',
 'charlax (Charles-AxelDein)',
 'ForrestKnight (ForrestKnight)',
 'revery-ui (revery-ui)',
 'alibaba (Alibaba)',
 'Microsoft (Microsoft)',
 'github (GitHub)',
 'facebook (Facebook)',
 'boazsegev (Bo)',
 'google (Google)',
 'cloudfetch',
 'sindresorhus (SindreSorhus)',
 'tensorflow',
 'apache (TheApacheSoftwareFoundation)',
 'DevonCrawford (DevonCrawford)',
 'ARMmbed (ArmMbed)',
 'vuejs (vuejs)',
 'fastai (fast.ai)',
 'QiShaoXuan (Qi)',
 'joelparkerhenderson (JoelParkerHenderson)',
 'torvalds (LinusTorvalds)',
 'CyC2018',
 'komeiji-satori (神楽坂覚々)',
 'script-8']
 ```

In [4]:
#your code

soup = bs(req, 'html.parser')
print(soup)


<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="utf-8"/>
<link href="https://github.githubassets.com" rel="dns-prefetch"/>
<link href="https://avatars.githubusercontent.com" rel="dns-prefetch"/>
<link href="https://github-cloud.s3.amazonaws.com" rel="dns-prefetch"/>
<link href="https://user-images.githubusercontent.com/" rel="dns-prefetch"/>
<link crossorigin="anonymous" href="https://github.githubassets.com/assets/frameworks-93d34cfdadb1618eb00917161bb7f744.css" integrity="sha512-k9NM/a2xYY6wCRcWG7f3ROm4X5CJNikViGX0N8YIxs6sUYAe/j08/RSHXr3fA9wLIy87AMFCgXm6jbvhZhIXWw==" media="all" rel="stylesheet">
<link crossorigin="anonymous" href="https://github.githubassets.com/assets/site-6547fa2bebd0a8c63e461573691982cb.css" integrity="sha512-ZUf6K+vQqMY+RhVzaRmCy2ePbSZad4TkaGRbd6v5gFt6f9Q/nqjDkBDjQgXNmZw7J9mcYxlsE4fhRw7CTluRow==" media="all" rel="stylesheet">
<link crossorigin="anonymous" href="https://github.githubassets.com/assets/behaviors-9a6d928a0cc4b9d03dc52f27ca22afa4.css" integr

In [5]:
devs = soup.findAll('h1', {'class':'h3 lh-condensed'})
devs[0].string

#list_names = [devs[name].string for name in range(len(devs))]
#list_names

In [6]:
names = [dev.get_text() for dev in devs]
names

['\n\n            Michele Locati\n ',
 '\n\n            Han Xiao\n ',
 '\n\n            Alon Zakai\n ',
 '\n\n            Ha Thach\n ',
 '\n\n            Joel Arvidsson\n ',
 '\n\n            Major Hayden\n ',
 '\n\n            Ahmet Alp Balkan\n ',
 '\n\n            Don Jayamanne\n ',
 '\n\n            Alexey Golub\n ',
 '\n\n            Diego Muracciole\n ',
 '\n\n            Simen Bekkhus\n ',
 '\n\n            Dan Davison\n ',
 '\n\n            Mo Gorhom\n ',
 '\n\n            José Padilla\n ',
 '\n\n            Yevgeniy Brikman\n ',
 '\n\n            Thibault Duplessis\n ',
 '\n\n            Nelson Osacky\n ',
 '\n\n            Gleb Bahmutov\n ',
 '\n\n            Soledad Galli\n ',
 '\n\n            Gerasimos (Makis) Maropoulos\n ',
 '\n\n            Nicolas P. Rougier\n ',
 '\n\n            snipe\n ',
 '\n\n            Anthony Sottile\n ',
 '\n\n            Julius Härtl\n ',
 '\n\n            Samuel Colvin\n ']

In [12]:
names = [name.replace('\n','') for name in names]
names

['            Michele Locati ',
 '            Han Xiao ',
 '            Alon Zakai ',
 '            Ha Thach ',
 '            Joel Arvidsson ',
 '            Major Hayden ',
 '            Ahmet Alp Balkan ',
 '            Don Jayamanne ',
 '            Alexey Golub ',
 '            Diego Muracciole ',
 '            Simen Bekkhus ',
 '            Dan Davison ',
 '            Mo Gorhom ',
 '            José Padilla ',
 '            Yevgeniy Brikman ',
 '            Thibault Duplessis ',
 '            Nelson Osacky ',
 '            Gleb Bahmutov ',
 '            Soledad Galli ',
 '            Gerasimos (Makis) Maropoulos ',
 '            Nicolas P. Rougier ',
 '            snipe ',
 '            Anthony Sottile ',
 '            Julius Härtl ',
 '            Samuel Colvin ']

In [8]:
clean = [clean.replace('\n','') for clean in names]
clean

['            Michele Locati ',
 '            Han Xiao ',
 '            Alon Zakai ',
 '            Ha Thach ',
 '            Joel Arvidsson ',
 '            Major Hayden ',
 '            Ahmet Alp Balkan ',
 '            Don Jayamanne ',
 '            Alexey Golub ',
 '            Diego Muracciole ',
 '            Simen Bekkhus ',
 '            Dan Davison ',
 '            Mo Gorhom ',
 '            José Padilla ',
 '            Yevgeniy Brikman ',
 '            Thibault Duplessis ',
 '            Nelson Osacky ',
 '            Gleb Bahmutov ',
 '            Soledad Galli ',
 '            Gerasimos (Makis) Maropoulos ',
 '            Nicolas P. Rougier ',
 '            snipe ',
 '            Anthony Sottile ',
 '            Julius Härtl ',
 '            Samuel Colvin ']

In [9]:
list_names = [i.replace('  ', '') for i in clean ]
list_names

['Michele Locati ',
 'Han Xiao ',
 'Alon Zakai ',
 'Ha Thach ',
 'Joel Arvidsson ',
 'Major Hayden ',
 'Ahmet Alp Balkan ',
 'Don Jayamanne ',
 'Alexey Golub ',
 'Diego Muracciole ',
 'Simen Bekkhus ',
 'Dan Davison ',
 'Mo Gorhom ',
 'José Padilla ',
 'Yevgeniy Brikman ',
 'Thibault Duplessis ',
 'Nelson Osacky ',
 'Gleb Bahmutov ',
 'Soledad Galli ',
 'Gerasimos (Makis) Maropoulos ',
 'Nicolas P. Rougier ',
 'snipe ',
 'Anthony Sottile ',
 'Julius Härtl ',
 'Samuel Colvin ']

#### Display the trending Python repositories in GitHub

The steps to solve this problem is similar to the previous one except that you need to find out the repository names instead of developer names.

In [10]:
# This is the url you will scrape in this exercise
url2 = 'https://github.com/trending/python?since=daily'

In [17]:
#your code
html2 = requests.get(url2).content
soup = bs(html2, 'lxml')

repo = soup.findAll('h1', {'class':'h3 lh-condensed'})
list_repo = [i.text for i in repo]
list_repo


['\n\n\n\n        jina-ai /\n\n      jina\n ',
 '\n\n\n\n        MTK-bypass /\n\n      bypass_utility\n ',
 '\n\n\n\n        ansible /\n\n      awx\n ',
 '\n\n\n\n        Asabeneh /\n\n      30-Days-Of-Python\n ',
 '\n\n\n\n        PaddlePaddle /\n\n      PaddleOCR\n ',
 '\n\n\n\n        RyanElliott10 /\n\n      wsbtickerbot\n ',
 '\n\n\n\n        facebookresearch /\n\n      detectron2\n ',
 '\n\n\n\n        public-apis /\n\n      public-apis\n ',
 '\n\n\n\n        huggingface /\n\n      transformers\n ',
 '\n\n\n\n        OWASP /\n\n      CheatSheetSeries\n ',
 '\n\n\n\n        demisto /\n\n      content\n ',
 '\n\n\n\n        pandas-dev /\n\n      pandas\n ',
 '\n\n\n\n        facebookresearch /\n\n      ParlAI\n ',
 '\n\n\n\n        facebookresearch /\n\n      detr\n ',
 '\n\n\n\n        pytorch /\n\n      vision\n ',
 '\n\n\n\n        wagtail /\n\n      wagtail\n ',
 '\n\n\n\n        great-expectations /\n\n      great_expectations\n ',
 '\n\n\n\n        sanic-org /\n\n      sanic\

In [23]:
clean_repo = [list_repo.replace('\n','') for list_repo in list_repo]
clean_repo

['        jina-ai /      jina ',
 '        MTK-bypass /      bypass_utility ',
 '        ansible /      awx ',
 '        Asabeneh /      30-Days-Of-Python ',
 '        PaddlePaddle /      PaddleOCR ',
 '        RyanElliott10 /      wsbtickerbot ',
 '        facebookresearch /      detectron2 ',
 '        public-apis /      public-apis ',
 '        huggingface /      transformers ',
 '        OWASP /      CheatSheetSeries ',
 '        demisto /      content ',
 '        pandas-dev /      pandas ',
 '        facebookresearch /      ParlAI ',
 '        facebookresearch /      detr ',
 '        pytorch /      vision ',
 '        wagtail /      wagtail ',
 '        great-expectations /      great_expectations ',
 '        sanic-org /      sanic ',
 '        pytorch /      fairseq ',
 '        mingrammer /      diagrams ',
 '        nodejs /      node-gyp ',
 '        deepfakes /      faceswap ',
 '        Vonng /      ddia ',
 '        spulec /      moto ',
 '        samuelcolvin /      pyd

In [21]:
repo_ready = [i.replace('    ', '') for i in clean_repo]
repo_ready

['jina-ai /  jina ',
 'MTK-bypass /  bypass_utility ',
 'ansible /  awx ',
 'Asabeneh /  30-Days-Of-Python ',
 'PaddlePaddle /  PaddleOCR ',
 'RyanElliott10 /  wsbtickerbot ',
 'facebookresearch /  detectron2 ',
 'public-apis /  public-apis ',
 'huggingface /  transformers ',
 'OWASP /  CheatSheetSeries ',
 'demisto /  content ',
 'pandas-dev /  pandas ',
 'facebookresearch /  ParlAI ',
 'facebookresearch /  detr ',
 'pytorch /  vision ',
 'wagtail /  wagtail ',
 'great-expectations /  great_expectations ',
 'sanic-org /  sanic ',
 'pytorch /  fairseq ',
 'mingrammer /  diagrams ',
 'nodejs /  node-gyp ',
 'deepfakes /  faceswap ',
 'Vonng /  ddia ',
 'spulec /  moto ',
 'samuelcolvin /  pydantic ']

In [27]:
repos = [repo_ready.replace('\c','') for repo_ready in repo_ready]
repos

['jina-ai /  jina ',
 'MTK-bypass /  bypass_utility ',
 'ansible /  awx ',
 'Asabeneh /  30-Days-Of-Python ',
 'PaddlePaddle /  PaddleOCR ',
 'RyanElliott10 /  wsbtickerbot ',
 'facebookresearch /  detectron2 ',
 'public-apis /  public-apis ',
 'huggingface /  transformers ',
 'OWASP /  CheatSheetSeries ',
 'demisto /  content ',
 'pandas-dev /  pandas ',
 'facebookresearch /  ParlAI ',
 'facebookresearch /  detr ',
 'pytorch /  vision ',
 'wagtail /  wagtail ',
 'great-expectations /  great_expectations ',
 'sanic-org /  sanic ',
 'pytorch /  fairseq ',
 'mingrammer /  diagrams ',
 'nodejs /  node-gyp ',
 'deepfakes /  faceswap ',
 'Vonng /  ddia ',
 'spulec /  moto ',
 'samuelcolvin /  pydantic ']

#### Display all the image links from Walt Disney wikipedia page

In [28]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/Walt_Disney'

In [29]:
#your code
html = requests.get(url).content
soup = bs(html, 'lxml')

In [31]:
link1 = []
for i in soup.findAll('img', {'src':re.compile('.JPG')}):
    link1.append(i.get('src'))
    
link1

['//upload.wikimedia.org/wikipedia/commons/thumb/d/df/Walt_Disney_1946.JPG/220px-Walt_Disney_1946.JPG',
 '//upload.wikimedia.org/wikipedia/commons/thumb/1/1a/Walt_Disney_Grave.JPG/170px-Walt_Disney_Grave.JPG',
 '//upload.wikimedia.org/wikipedia/commons/thumb/a/a9/Disney_Display_Case.JPG/170px-Disney_Display_Case.JPG']

In [32]:
link2 = []
for j in soup.findAll('img', {'src':re.compile('.jpg')}):
    link2.append(j.get('src'))
    
link2

['//upload.wikimedia.org/wikipedia/commons/thumb/c/c4/Walt_Disney_envelope_ca._1921.jpg/220px-Walt_Disney_envelope_ca._1921.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Newman_Laugh-O-Gram_%281921%29.webm/220px-seek%3D2-Newman_Laugh-O-Gram_%281921%29.webm.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Trolley_Troubles_poster.jpg/170px-Trolley_Troubles_poster.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/7/71/Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_of_Review_Magazine.jpg/170px-Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_of_Review_Magazine.jpg',
 '//upload.wikimedia.org/wikipedia/en/thumb/4/4e/Steamboat-willie.jpg/170px-Steamboat-willie.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/5/57/Walt_Disney_1935.jpg/170px-Walt_Disney_1935.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/c/cd/Walt_Disney_Snow_white_1937_trailer_screenshot_%2813%29.jpg/220px-Walt_Disney_Snow_white_1937_

In [33]:
link3 = []
link4 = []
for a in link1:
    link3.append('https://' + a)

for b in link2:
    link4.append('https://' + b)
    
link3+link4

['https:////upload.wikimedia.org/wikipedia/commons/thumb/d/df/Walt_Disney_1946.JPG/220px-Walt_Disney_1946.JPG',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/1/1a/Walt_Disney_Grave.JPG/170px-Walt_Disney_Grave.JPG',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/a/a9/Disney_Display_Case.JPG/170px-Disney_Display_Case.JPG',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/c/c4/Walt_Disney_envelope_ca._1921.jpg/220px-Walt_Disney_envelope_ca._1921.jpg',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Newman_Laugh-O-Gram_%281921%29.webm/220px-seek%3D2-Newman_Laugh-O-Gram_%281921%29.webm.jpg',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Trolley_Troubles_poster.jpg/170px-Trolley_Troubles_poster.jpg',
 'https:////upload.wikimedia.org/wikipedia/commons/thumb/7/71/Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_of_Review_Magazine.jpg/170px-Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_

#### Retrieve an arbitary Wikipedia page of "Python" and create a list of links on that page

In [34]:
# This is the url you will scrape in this exercise
url ='https://en.wikipedia.org/wiki/Python' 

In [35]:
#your code
html = requests.get(url).content
soup = bs(html, 'lxml')

In [36]:
lst1 = [i.text for i in soup.findAll('a')]
lst1

['',
 'Jump to navigation',
 'Jump to search',
 'Python',
 'python',
 'Pythons',
 'Python (genus)',
 '1 Computing',
 '2 People',
 '3 Roller coasters',
 '4 Vehicles',
 '5 Weaponry',
 '6 Other uses',
 '7 See also',
 'edit',
 'Python (programming language)',
 'CMU Common Lisp',
 'PERQ 3',
 'edit',
 'Python of Aenus',
 'Python (painter)',
 'Python of Byzantium',
 'Python of Catana',
 'Python Anghelo',
 'edit',
 'Python (Efteling)',
 'Python (Busch Gardens Tampa Bay)',
 'Python (Coney Island, Cincinnati, Ohio)',
 'edit',
 'Python (automobile maker)',
 'Python (Ford prototype)',
 'edit',
 'Python (missile)',
 'Python (nuclear primary)',
 'Colt Python',
 'edit',
 'PYTHON',
 'Python (film)',
 'Python (mythology)',
 'Monty Python',
 'Python (Monty) Pictures',
 'edit',
 'Cython',
 'Pyton',
 'Pithon',
 '',
 'disambiguation',
 'internal link',
 'https://en.wikipedia.org/w/index.php?title=Python&oldid=997582414',
 'Categories',
 'Disambiguation pages',
 'Human name disambiguation pages',
 'Disambig

In [37]:
lst2 = []

for j in lst1:
    if j != '':
        lst2.append(j)
        
lst2

['Jump to navigation',
 'Jump to search',
 'Python',
 'python',
 'Pythons',
 'Python (genus)',
 '1 Computing',
 '2 People',
 '3 Roller coasters',
 '4 Vehicles',
 '5 Weaponry',
 '6 Other uses',
 '7 See also',
 'edit',
 'Python (programming language)',
 'CMU Common Lisp',
 'PERQ 3',
 'edit',
 'Python of Aenus',
 'Python (painter)',
 'Python of Byzantium',
 'Python of Catana',
 'Python Anghelo',
 'edit',
 'Python (Efteling)',
 'Python (Busch Gardens Tampa Bay)',
 'Python (Coney Island, Cincinnati, Ohio)',
 'edit',
 'Python (automobile maker)',
 'Python (Ford prototype)',
 'edit',
 'Python (missile)',
 'Python (nuclear primary)',
 'Colt Python',
 'edit',
 'PYTHON',
 'Python (film)',
 'Python (mythology)',
 'Monty Python',
 'Python (Monty) Pictures',
 'edit',
 'Cython',
 'Pyton',
 'Pithon',
 'disambiguation',
 'internal link',
 'https://en.wikipedia.org/w/index.php?title=Python&oldid=997582414',
 'Categories',
 'Disambiguation pages',
 'Human name disambiguation pages',
 'Disambiguation pag

#### Number of Titles that have changed in the United States Code since its last release point 

In [38]:
# This is the url you will scrape in this exercise
url = 'http://uscode.house.gov/download/download.shtml'

In [39]:
#your code
html = requests.get(url).content
soup = bs(html, 'lxml')

titles = soup.findAll('div', class_=['usctitle', 'usctitlechanged'])[2::]
titles

[<div class="usctitle" id="us/usc/t1">
 
           Title 1 - General Provisions <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>,
 <div class="usctitle" id="us/usc/t2">
 
           Title 2 - The Congress
 
         </div>,
 <div class="usctitle" id="us/usc/t3">
 
           Title 3 - The President <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>,
 <div class="usctitle" id="us/usc/t4">
 
           Title 4 - Flag and Seal, Seat of Government, and the States <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>,
 <div class="usctitle" id="us/usc/t5">
 
           Title 5 - Government Organization and Employees <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>,
 <div class="usctitlechanged" id="us/usc/t6">
 
           Title 6 - Domestic Security
 
         </div>,
 <div class="usctitle" id="us/usc/t7">
 
           Title 7 - Agriculture
 
         </div>,
 <div class="usctitle" id="us/usc/t8">
 
           Title 

In [40]:
len(titles)

54

#### A Python list with the top ten FBI's Most Wanted names 

In [41]:
# This is the url you will scrape in this exercise
url = 'https://www.fbi.gov/wanted/topten'

In [43]:
#your code 
html = requests.get(url).content
soup = bs(html, 'lxml')

fbi = [i.text for i in soup.findAll('h3', class_='title')]
fbi

['\nALEJANDRO ROSALES CASTILLO\n',
 '\nARNOLDO JIMENEZ\n',
 '\nJASON DEREK BROWN\n',
 '\nALEXIS FLORES\n',
 '\nJOSE RODOLFO VILLARREAL-HERNANDEZ\n',
 '\nEUGENE PALMER\n',
 '\nRAFAEL CARO-QUINTERO\n',
 '\nROBERT WILLIAM FISHER\n',
 '\nBHADRESHKUMAR CHETANBHAI PATEL\n',
 '\nYASER ABDEL SAID\n']

In [45]:
fbi1 = []

for i in fbi:
    fbi1.append(i.split('\n')[1])
    
fbi1

['ALEJANDRO ROSALES CASTILLO',
 'ARNOLDO JIMENEZ',
 'JASON DEREK BROWN',
 'ALEXIS FLORES',
 'JOSE RODOLFO VILLARREAL-HERNANDEZ',
 'EUGENE PALMER',
 'RAFAEL CARO-QUINTERO',
 'ROBERT WILLIAM FISHER',
 'BHADRESHKUMAR CHETANBHAI PATEL',
 'YASER ABDEL SAID']

####  20 latest earthquakes info (date, time, latitude, longitude and region name) by the EMSC as a pandas dataframe

In [46]:
# This is the url you will scrape in this exercise
url = 'https://www.emsc-csem.org/Earthquake/'

In [48]:
#your code
html = requests.get(url).content
soup = bs(html, 'lxml')

In [72]:
info_date = soup.findAll('td', {'class':'tabev6'})[:20]
info_lat = soup.findAll('td', {'class':'tabev1'})[:40]
info_reg = soup.findAll('td', {'class':'tb_region'})[:20]

In [73]:
lat_long = [i.text.strip().split('\n') for i in info_lat]

lat = lat_long[0::2]
long = lat_long[1::2]

print(lat, long)

[['37.83'], ['54.23'], ['37.22'], ['41.27'], ['37.52'], ['32.40'], ['45.68'], ['0.26'], ['37.21'], ['39.01'], ['19.22'], ['37.18'], ['44.29'], ['38.17'], ['37.20'], ['38.17'], ['37.21'], ['37.19'], ['37.22'], ['37.21']] [['21.12'], ['159.67'], ['3.73'], ['75.39'], ['30.84'], ['67.52'], ['14.26'], ['99.35'], ['3.73'], ['30.42'], ['155.38'], ['3.71'], ['115.15'], ['117.97'], ['3.72'], ['117.82'], ['3.71'], ['3.73'], ['3.73'], ['3.73']]


In [74]:
region = [j.get_text() for i in info_reg]
region

AttributeError: 'str' object has no attribute 'get_text'

In [75]:
datetime = [k.text.strip().split('\n') for k in info_date]

datetime

[['earthquake2021-01-29\xa0\xa0\xa000:23:33.327min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa000:16:06.835min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa000:12:06.439min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa000:01:39.749min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa000:01:04.850min ago'],
 ['earthquake2021-01-29\xa0\xa0\xa000:00:01.051min ago'],
 ['earthquake2021-01-28\xa0\xa0\xa023:46:55.01hr 04min ago'],
 ['earthquake2021-01-28\xa0\xa0\xa023:45:32.01hr 06min ago'],
 ['earthquake2021-01-28\xa0\xa0\xa023:41:14.61hr 10min ago'],
 ['earthquake2021-01-28\xa0\xa0\xa023:38:52.61hr 12min ago'],
 ['earthquake2021-01-28\xa0\xa0\xa023:37:35.31hr 13min ago'],
 ['earthquake2021-01-28\xa0\xa0\xa023:32:25.21hr 19min ago'],
 ['earthquake2021-01-28\xa0\xa0\xa023:27:50.21hr 23min ago'],
 ['earthquake2021-01-28\xa0\xa0\xa023:25:08.41hr 26min ago'],
 ['earthquake2021-01-28\xa0\xa0\xa023:24:28.01hr 27min ago'],
 ['earthquake2021-01-28\xa0\xa0\xa022:57:36.21hr 53min ago'],
 ['earthquake2021-01-28\xa0\xa0\

In [76]:
date = []
time = []

for a in range(len(datetime)):
    path_date = '[0-9]+-[0-9]+-[0-9]+'
    path_time = '[0-9]+:[0-9]+:[0-9]+'
    date.extend([re.findall(path_date, a) for a in datetime[a]])
    time.extend([re.findall(path_time, a) for a in datetime[a]])
    
columns = [date, time, lat, long]
col = [[a for [a] in b] for b in columns]

In [79]:
earthquake = pd.DataFrame(list(zip(col[0], col[1], col[2], col[3])),
                         columns = ['date','time','lat','long'])[:20]

In [80]:
earthquake

date      time    lat    long
0   2021-01-29  00:23:33  37.83   21.12
1   2021-01-29  00:16:06  54.23  159.67
2   2021-01-29  00:12:06  37.22    3.73
3   2021-01-29  00:01:39  41.27   75.39
4   2021-01-29  00:01:04  37.52   30.84
5   2021-01-29  00:00:01  32.40   67.52
6   2021-01-28  23:46:55  45.68   14.26
7   2021-01-28  23:45:32   0.26   99.35
8   2021-01-28  23:41:14  37.21    3.73
9   2021-01-28  23:38:52  39.01   30.42
10  2021-01-28  23:37:35  19.22  155.38
11  2021-01-28  23:32:25  37.18    3.71
12  2021-01-28  23:27:50  44.29  115.15
13  2021-01-28  23:25:08  38.17  117.97
14  2021-01-28  23:24:28  37.20    3.72
15  2021-01-28  22:57:36  38.17  117.82
16  2021-01-28  22:52:57  37.21    3.71
17  2021-01-28  22:48:31  37.19    3.73
18  2021-01-28  22:47:26  37.22    3.73
19  2021-01-28  22:45:42  37.21    3.73

#### Display the date, days, title, city, country of next 25 hackathon events as a Pandas dataframe table

In [ ]:
# This is the url you will scrape in this exercise
url ='https://hackevents.co/hackathons'

In [ ]:
#your code

#### Count number of tweets by a given Twitter account.

You will need to include a ***try/except block*** for account names not found. 
<br>***Hint:*** the program should count the number of tweets for any provided account

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
#your code

#### Number of followers of a given twitter account

You will need to include a ***try/except block*** in case account/s name not found. 
<br>***Hint:*** the program should count the followers for any provided account

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
#your code

#### List all language names and number of related articles in the order they appear in wikipedia.org

In [81]:
# This is the url you will scrape in this exercise
url = 'https://www.wikipedia.org/'

In [82]:
#your code
html = requests.get(url).content
soup = bs(html, 'lxml')

In [83]:
wiki = [i.text for i in soup.findAll('div', class_='central-featured-lang')]
wiki

['\n\nEnglish\n6\xa0235\xa0000+ articles\n\n',
 '\n\n日本語\n1\xa0250\xa0000+ 記事\n\n',
 '\n\nDeutsch\n2\xa0528\xa0000+ Artikel\n\n',
 '\n\nEspañol\n1\xa0655\xa0000+ artículos\n\n',
 '\n\nРусский\n1\xa0693\xa0000+ статей\n\n',
 '\n\nFrançais\n2\xa0292\xa0000+ articles\n\n',
 '\n\nItaliano\n1\xa0669\xa0000+ voci\n\n',
 '\n\n中文\n1\xa0172\xa0000+ 條目\n\n',
 '\n\nPolski\n1\xa0452\xa0000+ haseł\n\n',
 '\n\nPortuguês\n1\xa0053\xa0000+ artigos\n\n']

In [85]:
wiki_list = []
for i in wiki:
    wiki_list.append(i.split("\n\n")[1])

wiki_list

['English\n6\xa0235\xa0000+ articles',
 '日本語\n1\xa0250\xa0000+ 記事',
 'Deutsch\n2\xa0528\xa0000+ Artikel',
 'Español\n1\xa0655\xa0000+ artículos',
 'Русский\n1\xa0693\xa0000+ статей',
 'Français\n2\xa0292\xa0000+ articles',
 'Italiano\n1\xa0669\xa0000+ voci',
 '中文\n1\xa0172\xa0000+ 條目',
 'Polski\n1\xa0452\xa0000+ haseł',
 'Português\n1\xa0053\xa0000+ artigos']

In [101]:
wiki_clean = []
for j in wiki_list:
    path_1 = '日本語, 中文, Русский, Français, Português'
    wiki_clean.append(re.findall(path_1, j)[4])
    
wiki_clean

IndexError: list index out of range

In [95]:
wiki_trash = [k.text for k in soup.findAll('bdi', dir='ltr')]
wiki_trash

['6\xa0235\xa0000+',
 '1\xa0250\xa0000+',
 '2\xa0528\xa0000+',
 '1\xa0655\xa0000+',
 '1\xa0693\xa0000+',
 '2\xa0292\xa0000+',
 '1\xa0669\xa0000+',
 '1\xa0172\xa0000+',
 '1\xa0452\xa0000+',
 '1\xa0053\xa0000+',
 '\n1\xa0000\xa0000+\n',
 '\n100\xa0000+\n',
 '\n10\xa0000+\n',
 '\n1\xa0000+\n',
 '\n100+\n']

In [96]:
wiki_bag = []
for i in wiki_trash:
    path = '[^[\xa0]+]*'
    wiki_bag.append(re.findall(path, i))
    
wiki_bag

[['6', '235', '000+'],
 ['1', '250', '000+'],
 ['2', '528', '000+'],
 ['1', '655', '000+'],
 ['1', '693', '000+'],
 ['2', '292', '000+'],
 ['1', '669', '000+'],
 ['1', '172', '000+'],
 ['1', '452', '000+'],
 ['1', '053', '000+'],
 ['\n1', '000', '000+\n'],
 ['\n100', '000+\n'],
 ['\n10', '000+\n'],
 ['\n1', '000+\n'],
 ['\n100+\n']]

In [102]:
new_list = []
for i in new_list:
    a = ''
    for j in i:
        a += j
    new_list.append(a)

bag = new_list[:10]


new_list_2 = []
for i in range(len(wiki_clean)):
    bew_list_2.append(wiki_clean[i] + ' ' + bag[i])
    
new_list_2

[]

#### A list with the different kind of datasets available in data.gov.uk 

In [ ]:
# This is the url you will scrape in this exercise
url = 'https://data.gov.uk/'

In [ ]:
#your code 

#### Top 10 languages by number of native speakers stored in a Pandas Dataframe

In [ ]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/List_of_languages_by_number_of_native_speakers'

In [ ]:
#your code

### BONUS QUESTIONS

#### Scrape a certain number of tweets of a given Twitter account.

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
# your code

#### IMDB's Top 250 data (movie name, Initial release, director name and stars) as a pandas dataframe

In [ ]:
# This is the url you will scrape in this exercise 
url = 'https://www.imdb.com/chart/top'

In [ ]:
# your code

#### Movie name, year and a brief summary of the top 10 random movies (IMDB) as a pandas dataframe.

In [ ]:
#This is the url you will scrape in this exercise
url = 'http://www.imdb.com/chart/top'

In [ ]:
#your code

#### Find the live weather report (temperature, wind speed, description and weather) of a given city.

In [ ]:
#https://openweathermap.org/current
city = city=input('Enter the city:')
url = 'http://api.openweathermap.org/data/2.5/weather?'+'q='+city+'&APPID=b35975e18dc93725acb092f7272cc6b8&units=metric'

In [ ]:
# your code

#### Book name,price and stock availability as a pandas dataframe.

In [ ]:
# This is the url you will scrape in this exercise. 
# It is a fictional bookstore created to be scraped. 
url = 'http://books.toscrape.com/'

In [ ]:
#your code